squirrel.frontier/src/main/java/org/dice_research/squirrel/data/uri/norm/NormalizerImpl.java

```diff
@@ -20,122 +24,218 @@
+    }
+
+    @Override
+    public CrawleableUri normalize(CrawleableUri uri) {
+        URI uriObject = uri.getUri();
+        boolean changed = false;
+        // normalize path
+        String path = uriObject.getRawPath();
+        if (path != null) {
+            String temp = normalizePath(path);
+            if (temp != path) {
+                path = temp;
+                changed = true;
+            }
+        }
+
+
+        // Copy Normalization from
+        // https://github.com/crawler-commons/crawler-commons/blob/master/src/main/java/crawlercommons/filters/basic/BasicURLNormalizer.java
+        // OR use URI.normalize()
+
+        // Check whether the query part of a URI has to be sorted
+        String query = uriObject.getQuery();
+        if(query != null){
+            if(query.length() > 0) {
+                String[] queryList = query.split("&");
+                Arrays.sort(queryList);
+                List<String> queries = new ArrayList<>(Arrays.asList(queryList));
+                List<String> toRemove = new ArrayList<>();
+                for(String queryParameter : queries){
+                    //removing session ids
+                    if(sessionIDs.contains(queryParameter.split("=")[0].toLowerCase())){
+                        toRemove.add(queryParameter);
+                    }
+                }
+                queries.removeAll(toRemove);
+                query = String.join("&", queries);
+                changed = true;
+            }
+            else{
+                query = null;
+                changed = true;
+            }
+        }
+
+        //Remove default ports
+        int port = uriObject.getPort();
+        String scheme = uriObject.getScheme();
+        if(port != -1){
+            if(defaultPortMap.containsKey(scheme)){
+                if(port == defaultPortMap.get(scheme)){
+                    port = -1;
+                    changed = true;
+                }
+            }
+        }
+        // Filter fragments (i.e., delete them)
+        String fragment = uriObject.getFragment();
+        if ((fragment != null) && (fragment.length() > 0)) {
+            changed = true;
+        }
+
+        // convert host and scheme to lower case
+        String host = uriObject.getHost();
+        if(!scheme.equals(scheme.toLowerCase()) || !host.equals(host.toLowerCase())){
+            scheme = scheme.toLowerCase();
+            host = host.toLowerCase();
+            changed = true;
+        }
+
+        // Filter attributes of the URI
+        // uriObject.getQuery();
+
+        if (changed) {
+            // create new URI object;
+            URIBuilder builder = new URIBuilder(uriObject);
+            builder.setFragment(null);
+            builder.setPath(path);
+            builder.setCustomQuery(query);
+            builder.setPort(port);
+            builder.setHost(host);
+            builder.setScheme(scheme);
+            try {
+                uri = new CrawleableUri(builder.build());
```

Comment:
This is not sufficient. Think of all the data that might be attached to the URI which will be lost by simply overriding the object in this way :open_mouth:

Reply:
The error already existed in the original code. Sorry for that. The attached data should be copied to the new URI object.